app/src/dex/router/fill_route2.rs

```diff
@@ -0,0 +1,197 @@
+use std::collections::{BTreeMap, HashMap};
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use futures::{StreamExt, TryStreamExt};
+use penumbra_crypto::{asset, dex::DirectedTradingPair, fixpoint::U128x128, Amount, Value};
+use penumbra_storage::StateWrite;
+use tracing::instrument;
+
+use crate::dex::{PositionManager, PositionRead};
+
+#[async_trait]
+pub trait FillRoute2: StateWrite {
+    ///
+    ///
+    /// # Invariants
+    ///
+    /// It is an error to call `fill_route` on a route that does not have at least one position for each hop.
+    #[instrument(skip(self, input, hops, spill_price))]
+    async fn fill_route(
+        &mut self,
+        mut input: Value,
+        hops: &[asset::Id],
+        spill_price: Option<U128x128>,
+    ) -> Result<(Value, Value)> {
+        // Switch from representing hops implicitly as a sequence of asset IDs to
+        // representing them explicitly as a sequence of directed trading pairs.
+        let route = std::iter::once(input.asset_id)
+            .chain(hops.iter().cloned())
+            .collect::<Vec<_>>();
+        // Break down the route into a sequence of pairs to visit.
+        let pairs = breakdown_route(&route)?;
+
+        // We want to ensure that any particular position is used at most once over the route,
+        // even if the route has cycles at the macro-scale. To do this, we store the streams
+        // of positions for each pair, taking care to only construct one stream per distinct pair.
+        let mut positions_by_price = HashMap::new();
+        for pair in &pairs {
+            positions_by_price
+                .entry(pair.clone())
+                .or_insert_with(|| self.positions_by_price(&pair));
+        }
+
+        // Record a trace of the execution along the current route,
+        // starting with all-zero amounts.
+        let mut trace: Vec<Value> = std::iter::once(Value {
+            amount: 0u64.into(),
+            asset_id: input.asset_id,
+        })
+        .chain(hops.iter().map(|asset_id| Value {
+            amount: 0u64.into(),
+            asset_id: asset_id.clone(),
+        }))
+        .collect();
+
+        let mut output = Value {
+            amount: 0u64.into(),
+            asset_id: route
+                .last()
+                .cloned()
+                .ok_or_else(|| anyhow::anyhow!("called fill_route with empty route"))?,
+        };
+
+        // TODO: this does not allow traversing the same macro-edge in opposite directions
+        // (This impl will produce WRONG RESULTS if you try to do that!)
+        // later: implement fast path for non-opposite directions, slow path (check for duplicate position ids)
+
+        let mut frontier = Vec::new();
```

Comment:
I think we should change `frontier` to be a `Vec<(Position, position::Id)>`, saving the position IDs as well as the positions.  Then, before adding or replacing an existing position on the frontier, we should check whether its ID is already present in the frontier.

Reply:
Actually, this could just be a standalone BTreeSet, seems more efficient